planner/cascades/transformation_rules.go

```diff
@@ -981,6 +982,89 @@ func (r *MergeAdjacentProjection) OnTransform(old *memo.ExprIter) (newExprs []*m
 	return []*memo.GroupExpr{newProjExpr}, true, false, nil
 }
 
+// PushTopNDownOuterJoin pushes topN to outer join.
+type PushTopNDownOuterJoin struct {
+	baseRule
+}
+
+// NewRulePushTopNDownOuterJoin creates a new Transformation PushTopNDownOuterJoin.
+// The pattern of this rule is: `TopN -> Join`.
+func NewRulePushTopNDownOuterJoin() Transformation {
+	rule := &PushTopNDownOuterJoin{}
+	rule.pattern = memo.BuildPattern(
+		memo.OperandTopN,
+		memo.EngineTiDBOnly,
+		memo.NewPattern(memo.OperandJoin, memo.EngineTiDBOnly),
+	)
+	return rule
+}
+
+// Match implements Transformation interface.
+// Use appliedRuleSet in GroupExpr to avoid re-apply rules.
+func (r *PushTopNDownOuterJoin) Match(expr *memo.ExprIter) bool {
+	if expr.GetExpr().HasAppliedRule(r) {
+		return false
+	}
+	join := expr.Children[0].GetExpr().ExprNode.(*plannercore.LogicalJoin)
+	switch join.JoinType {
+	case plannercore.LeftOuterJoin, plannercore.LeftOuterSemiJoin, plannercore.AntiLeftOuterSemiJoin, plannercore.RightOuterJoin:
+		return true
+	default:
+		return false
+	}
+}
+
+func pushTopNDownOuterJoinToChild(topN *plannercore.LogicalTopN, outerGroup *memo.Group, innerGroup *memo.Group) *memo.Group {
```

Comment:
how about bind this function to struct `PushTopNDownOuterJoin`?

Reply:
I don't think it‘s quite different, any reason for that?